<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>M-SALSA by SALSA-W</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">M-SALSA</h1>
      <h2 class="project-tagline">M-SALSA (Sequence Alignment by Local Search Algorithm) is a local search algorithm for solving Multiple Sequence Alignment</h2>
      <a href="https://github.com/SALSA-W/M-SALSA" class="btn">View on GitHub</a>
      <a href="https://github.com/SALSA-W/M-SALSA/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/SALSA-W/M-SALSA/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="what-is-m-salsa" class="anchor" href="#what-is-m-salsa" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>What is M-SALSA?</h1>

<p><a href="https://en.wikipedia.org/wiki/Multiple_sequence_alignment">Multiple sequence alignment (MSA)</a> is an important problem in bioinformatics. Its purpose is to find relationships among residues of biological sequences connected by an evolutionary, structural or functional relationship.</p>

<p><strong>M-SALSA</strong> (Sequence Alignment by Local Search Algorithm) is a stochastic local search algorithm that improves alignments generated by other MSA programs.</p>

<p>Two versions of M-SALSA are available: one implemented using c++ and the other using java.
It is also available a web application that gives an easier user experience. It automatically calls an external program that calculates the initial alignment to be enhanced by M-SALSA. </p>

<h1>
<a id="how-to-install" class="anchor" href="#how-to-install" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>How to install</h1>

<p>For install instructions visit <a href="https://github.com/SALSA-W/M-SALSA/wiki/Install-instructions">M-SALSA wiki</a>.</p>

<h1>
<a id="contribution" class="anchor" href="#contribution" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Contribution</h1>

<p>To report a bug or suggest an enhancement visit <a href="https://github.com/SALSA-W/M-SALSA/issues">M-SALSA issue tracker</a>.</p>

<h1>
<a id="command-line-usage" class="anchor" href="#command-line-usage" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Command line usage</h1>

<h2>
<a id="c-version" class="anchor" href="#c-version" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>c++ version</h2>

<p>In order to use M-SALSA command line version on linux you must use this sintax:<br>
<code>M-SALSA [PARAMETERS]</code><br>
Parameters must be specified writing the name of the option followed by a space and the option value:<br>
<code>-optionName option_value</code><br>
An example of M-SALSA usage is:<br>
<code>M-SALSA -inputFile BB11001.tfa -outputFile BB11001_final.tfa -phTreeFile BB11001.dnd -scoringMatrix BLOSUM80</code><br>
</p>

<h3>
<a id="mandatory-parameters" class="anchor" href="#mandatory-parameters" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Mandatory parameters:</h3>

<dl>
<dt>-inputFile:</dt>       <dd>path of a file containing the initial alignment. The file must be in <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA format</a></dd>
<dt>-outputFile:</dt>      <dd>path of the output file. This will be in FASTA format as well</dd>
<dt>-phTreeFile:</dt>      <dd>file containing the guide tree, used by M-SALSA in order to generate correct weigths for the WSP-Score (link). The file must be in <a href="https://en.wikipedia.org/wiki/Newick_format">Newicks format</a></dd>
</dl>

<hr>

<h3>
<a id="optional-parameters" class="anchor" href="#optional-parameters" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Optional parameters:</h3>

<dl>
<dt>-GOP:</dt>            <dd>GAP Opening Penalty (<i>default 8</i>)</dd>
<dt>-GEP:</dt>            <dd>GAP Extension Penalty (<i>default 5</i>)</dd>
<dt>-gamma:</dt>          <dd>dimension of the range of positions a GAP can move in during an iteration (<i>default 30</i>)</dd>
<dt>-type:</dt>           <dd>type of sequences. Possible options are DNA, RNA and PROTEINS (default PROTEINS</i>)</dd>
<dt>-scoringMatrix:</dt>  <dd>scoring matrix (<i>default BLOSUM62</i>). Available matrices are: BLOSUM30, BLOSUM45, BLOSUM50, BLOSUM62, BLOSUM80, IUB, PAM20, PAM60, PAM120, PAM350.</dd>
<dt>-scoringMatrixPath:</dt>  <dd>scoring matrix file. For more information on scoring matrix file format visit <a href="https://github.com/SALSA-W/M-SALSA/wiki">M-SALSA wiki</a></dd>
<dt>-matrixSerie:</dt>    <dd>matrix serie. Possible options: BLOSUM or PAM (<i>default BLOSUM</i>)</dd>
<dt>-distanceMatrix:</dt> <dd>distance matrix file</dd>
<dt>-minIt:</dt>          <dd>minimum number of iterations (<i>default 1000</i>)</dd>
<dt>-pSplit:</dt>         <dd>probability of split (<i>default 0.1</i>)</dd>
<dt>-terminal:</dt>       <dd>the strategy to be used to manage terminal GAPs. Possible values:
                          <ul>
                              <li> ONLY_GEP: GOP=0 only for terminal GAPs (<i>default value</i>)</li>
                              <li> BOTH: both opening and extension penalty for terminal GAPs</li>
                          </ul></dd>
</dl>

<hr>

<h3>
<a id="substitution-matrix-choice" class="anchor" href="#substitution-matrix-choice" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Substitution matrix choice:</h3>

<p>If no optional parameters are provided M-SALSA default substitution matrix is BLOSUM62. If user specifies the type parameter, the default matrix is chosen according to it. For PROTEINS the default matrix is still BLOSUM62, for DNA it is IUB, for RNA no default matrix exists. Therefore, if input consists on RNA sequences user must provide a substitution matrix.</p>

<p>In case of PROTEINS type it is also possible to specify a distance matrix, a matrix containing the distance between each couple of sequences in the alignment. M-SALSA uses such a matrix to automatically infer the correct substitution matrix. The chosen matrix depends also an the matrixSerie parameter, that can be BLOSUM or PAM (if not specified it will use a BLOSUM matrix).</p>

<h2>
<a id="java-version" class="anchor" href="#java-version" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>java version</h2>

<h1><a id="documentation" class="anchor" href="#documentation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Documentation</h1>
<p>The documentation is available at <a href="http://salsa-w.github.io/M-SALSA/apidocs">http://salsa-w.github.io/M-SALSA/apidocs</a></p>
<p>More informations available on <a href="https://github.com/SALSA-W/M-SALSA/wiki/API-Docs">https://github.com/SALSA-W/M-SALSA/wiki/API-Docs</a></p>

<h1>
<a id="license" class="anchor" href="#license" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>License</h1>

<p>Copyright 2015 Alessandro Daniele, Fabio Cesarato, Andrea Giraldin</p>

<p>Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at</p>

<p><a href="http://www.apache.org/licenses/LICENSE-2.0">http://www.apache.org/licenses/LICENSE-2.0</a></p>

<p>Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.</p>

<p><a href="https://raw.githubusercontent.com/SALSA-W/M-SALSA/master/LICENSE"><img src="http://img.shields.io/badge/license-APACHE2-blue.svg" alt="License"></a></p>

<h1>
<a id="support-or-contact" class="anchor" href="#support-or-contact" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Support or Contact</h1>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/SALSA-W/M-SALSA">M-SALSA</a> is maintained by <a href="https://github.com/SALSA-W">SALSA-W</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
